In [4]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [6]:
import pandas

mydataset = {
    'cars': ["BMW", "Volvo", "Ford"],
    'passings': [3, 7, 2]
}

myVar = pandas.DataFrame(mydataset)

print(myVar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


In [7]:
import pandas as pd

mydataset = {
    'cars': ["BMW", "Volvo", "Ford"],
    'passings': [3, 7, 2]
}

myVar = pd.DataFrame(mydataset)

print(myVar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


Checking Pandas Version

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **بررسی نسخه پانداس نصب شده**

با نوشتن دستور version میتوان به نسخه‌ی پانداس نصب شده دسترسی پیدا کرد.


</section>

In [8]:
import pandas as pd

print(pd.__version__)

2.1.1


___
Checking Pandas Version

<section style="direction: rtl; text-align: justify; margin: 30px;">

## **سری پانداس**

سری‌های پانداس مانند یک ستون در یک جدول است.
این یک آرایه یک بعدی برای نگهداری اطلاعات از هر نوعی می‌باشد.

</section>

In [10]:
import pandas as pd

a = [1, 7, 2]

myVar = pd.Series(a)

print(myVar)

0    1
1    7
2    2
dtype: int64


Labels

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **برچسب‌ها**

اگر چیز دیگری مشخص نشده باشد، مقادیر با شماره شاخص خود برچسب گذاری می شوند. مقدار اول دارای شاخص 0، مقدار دوم دارای شاخص 1 و غیره است.
از این برچسب می توان برای دسترسی به یک مقدار مشخص استفاده کرد.


</section>

In [11]:
# Return the first value of the Series:

a = [1, 7, 2]

myVar = pd.Series(a)

print(myVar[0])

1


Create Labels

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **ساخت برچسب‌ها**

با آرگومان index، می توانید برچسب های خود را نام گذاری کنید.


</section>

In [12]:
# Create your own labels:

import pandas as pd

a = [1, 7, 2]

myVar = pd.Series(a, index=["x", "y", "z"])

print(myVar)

x    1
y    7
z    2
dtype: int64


In [14]:
# Return the value of "y":

import pandas as pd

a = [1, 7, 2]

myVar = pd.Series(a, index=["x", "y", "z"])

print(myVar["y"])

7


Key/Value Objects as Series

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **کلید/مقدار اشیاء به صورت سری**

همچنین می‌توانید از یک شی کلید/مقدار، مانند دیکشنری (dictionary)، هنگام ایجاد یک سری استفاده کنید.

</section>

In [15]:
# Create a simple Pandas Series from a dictionary:

calories = {"day1": 420, "day2": 380, "day3": 390}

myVar = pd.Series(calories)

print(myVar)

day1    420
day2    380
day3    390
dtype: int64


In [17]:
# Create a Series using only data from "day1" and "day2":

calories = {"day1": 420, "day2": 380, "day3": 390}

myVar = pd.Series(calories, index=["day1", "day2"])

print(myVar)

day1    420
day2    380
dtype: int64


DataFrames

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **دیتافریم**

مجموعه داده ها در پانداس معمولا جداول چند بعدی هستند که DataFrames نامیده می شوند.
سری مانند یک ستون است، یک DataFrame کل جدول است.

</section>

In [18]:
# Create a DataFrame from two Series:

import pandas as pd

data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

myVar = pd.DataFrame(data)

print(myVar)

   calories  duration
0       420        50
1       380        40
2       390        45


___
Pandas DataFrames

<section style="direction: rtl; text-align: justify; margin: 30px;">

## **دیتافریم‌های پانداس**

Pandas DataFrame یک ساختار داده دو بعدی است، مانند یک آرایه دو بعدی، یا یک جدول با ردیف و ستون.


</section>

In [21]:
import pandas as pd

data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

# load data into a DataFrame object:
df = pd.DataFrame(data)

print(df)

   calories  duration
0       420        50
1       380        40
2       390        45


Pandas DataFrames

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **پیدا کردن ردیف‌ها**

همانطور که از نتیجه بالا می بینید، DataFrame مانند یک جدول با ردیف و ستون است.
پانداس از ویژگی loc برای برگرداندن یک یا چند ردیف مشخص شده استفاده می کنند.

</section>

In [10]:
# Return row 0:

import pandas as pd

data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

df = pd.DataFrame(data)

print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


In [2]:
# Return row 0 and 1:

import pandas as pd

data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

df = pd.DataFrame(data)

print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


Named Indexes

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **شاخص‌های نام گذاری شده**

با آرگومان index، می توانید فهرست های خود را نام ببرید.


</section>

In [9]:
# Add a list of names to give each row a name:

import pandas as pd

data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index=["day1", "day2", "day3"])

print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        45


Locate Named Indexes

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **پیدا کردن شاخص‌های نام گذاری شده**

برای برگرداندن ردیف(های) مشخص شده از شاخص نامگذاری شده در ویژگی loc استفاده کنید.

</section>

In [13]:
# Return "day2":

import pandas as pd

data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index=["day1", "day2", "day3"])

print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


Load Files Into a DataFrame

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **بارگزاری فایل‌ها در یک دیتافریم**

اگر مجموعه داده‌های شما در یک فایل ذخیره می‌شوند، پانداس می‌تواند آنها را در یک DataFrame بارگذاری کند.

</section>

In [3]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


___
Pandas Read CSV

<section style="direction: rtl; text-align: justify; margin: 30px;">

## **خواندن فایل CSV در پانداس**

یک راه ساده برای ذخیره مجموعه داده های بزرگ استفاده از فایل های CSV (فایل های جدا شده با کاما) است.
فایل‌های CSV حاوی متن ساده است و فرمت شناخته شده‌ای است که برای همه از جمله کتابخانه پانداس قابل خواندن است.

</section>

In [4]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [5]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


max_rows

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **تمام ردیف‌ها**

تعداد ردیف های برگشتی در تنظیمات گزینه Pandas تعریف شده است.
می توانید حداکثر ردیف های سیستم خود را با عبارت pd.options.display.max_rows بررسی کنید.


</section>

In [6]:
import pandas as pd

print(pd.options.display.max_rows)

60


<section style="direction: rtl; text-align: justify; margin: 30px;">

در سیستم من این عدد 60 است، به این معنی که اگر DataFrame شامل بیش از 60 ردیف باشد، دستور print(df) فقط سرصفحه ها و 5 سطر اول و آخر را برمی گرداند.

شما می توانید حداکثر تعداد ردیف ها را با همان عبارت تغییر دهید.


</section>

In [12]:
import pandas as pd

pd.options.display.max_rows = 9999

df = pd.read_csv('data.csv')

print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

Pandas Read JSON

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **خواندن فایل‌های JSON**

مجموعه داده های بزرگ اغلب به صورت JSON ذخیره یا استخراج می شوند.
JSON متن ساده است، اما فرمت یک شی را دارد و در دنیای برنامه نویسی از جمله پانداها به خوبی شناخته شده است.
در مثال هایمان از یک فایل JSON به نام 'data.json' استفاده خواهیم کرد.



</section>

In [13]:
import pandas as pd

df = pd.read_json('data.js')

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

Dictionary as JSON

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **دیکشنری به عنوان JSON**

اگر کد JSON شما در یک فایل نیست، بلکه در یک دیکشنری پایتون است، می توانید آن را مستقیماً در یک DataFrame بارگذاری کنید:


</section>

In [14]:
import pandas as pd

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}

df = pd.DataFrame(data)

print(df)

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


___
Pandas - Analyzing DataFrames

<section style="direction: rtl; text-align: justify; margin: 30px;">

## **تجزیه و تحلیل پانداس**

یکی از پرکاربردترین روش ها برای دریافت نمای کلی از DataFrame، متد head() است.
متد head() سرصفحه ها و تعداد مشخصی از ردیف ها را که از بالا شروع می شود، برمی گرداند.

</section>

In [15]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.head(10))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


In [16]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


<section style="direction: rtl; text-align: justify; margin: 30px;">

همچنین یک متد tail() برای مشاهده آخرین ردیف های DataFrame وجود دارد.
متد tail() سرصفحه ها و تعداد مشخصی از ردیف ها را از پایین شروع می کند.

</section>

In [17]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.tail())

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


Info About the Data

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **اطلاعات داده‌ها**

شی DataFrames متدی به نام info() دارد که اطلاعات بیشتری در مورد مجموعه داده به شما می دهد.


</section>

In [18]:
import pandas as pd

df = pd.read_csv('data.csv')

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


Result Explained

<section style="direction: rtl; text-align: justify; margin: 30px;">

نتیجه به ما می‌گوید که داده ما شامل 169 ردیف و 4 ستون می‌باشد.

رنج داده‌های ما بین 0 تا 168 است.

متد info() همچنین به ما می گوید که در هر ستون چند مقدار Non-Null وجود دارد و در مجموعه داده های ما به نظر می رسد که 164 عدد از 169 مقدار Non-Null در ستون "Calories" وجود دارد.

به این معنی که به هر دلیلی در ستون "کالری" 5 ردیف بدون مقدار وجود دارد.

مقادیر خالی یا مقادیر Null هنگام تجزیه و تحلیل داده ها می توانند بد باشند، و شما باید ردیف هایی با مقادیر خالی را حذف کنید. این یک گام به سمت چیزی است که پاکسازی داده‌ها نامیده می‌شود و در فصل‌های بعدی بیشتر در مورد آن خواهیم آموخت.

</section>

___
Pandas - Cleaning Data

<section style="direction: rtl; text-align: justify; margin: 30px;">

## **پاکسازی داده‌ها**

پاکسازی داده ها به معنای اصلاح داده های بد در مجموعه داده شما است.

داده های بد می تواند:

- شامل سلول‌های خالی باشد.
- شامل اطلاعات با فرمت اشتباه باشد.
- شامل داده‌های اشتباه باشد.
- شامل داده‌های تکراری باشد.

در این آموزش یاد می گیرید که چگونه با همه آنها کنار بیایید.


</section>

___
Pandas - Cleaning Empty Cells

<section style="direction: rtl; text-align: justify; margin: 30px;">

## **پاکسازی سلول‌های خالی**

سلول‌های خالی به طور بالقوه می‌توانند هنگام تجزیه و تحلیل داده‌ها، نتیجه اشتباهی به شما بدهند.


### **حذف ردیف‌های خالی**

یکی از راه های مقابله با سلول های خالی حذف ردیف هایی است که حاوی سلول های خالی هستند.
این معمولاً مشکلی ندارد، زیرا مجموعه داده‌ها می‌توانند بسیار بزرگ باشند و حذف چند ردیف تأثیر زیادی در نتیجه نخواهد داشت.

</section>

In [19]:
import pandas as pd

df = pd.read_csv('data.csv')

new_df = df.dropna()

print(new_df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

<section style="direction: rtl; text-align: justify; margin: 30px;">

توجه: به طور پیش‌فرض، متد dropna() یک DataFrame جدید برمی‌گرداند و نسخه اصلی را تغییر نخواهد داد.

اگر می خواهید DataFrame اصلی را تغییر دهید، از آرگومان inplace = True استفاده کنید:


</section>

In [20]:
import pandas as pd

df = pd.read_csv('data.csv')

df.dropna(inplace=True)

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45    105       132     246.0
24         60   

<section style="direction: rtl; text-align: justify; margin: 30px;">

توجه: اکنون dropna (inplace = True) DataFrame جدیدی را برنمی‌گرداند، اما تمام ردیف‌های حاوی مقادیر NULL را از DataFrame اصلی حذف می‌کند.

</section>

Replace Empty Values

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **جایگذاری مقادیر خالی**

روش دیگر برای مقابله با سلول های خالی، درج یک مقدار جدید به جای آن است.
به این ترتیب شما مجبور نیستید تمام سطرها را فقط به دلیل چند سلول خالی حذف کنید.
متد fillna() به ما اجازه می دهد تا سلول های خالی را با یک مقدار جایگزین کنیم:

</section>

In [21]:
import pandas as pd

df = pd.read_csv('data.csv')

df.fillna(130, inplace=True)

Replace Only For Specified Columns

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **فقط برای ستون های مشخص شده جایگزین کنید**

مثال بالا جایگزین تمام سلول های خالی در کل Data Frame می شود.
فقط برای جایگزینی مقادیر خالی برای یک ستون، نام ستون را برای DataFrame مشخص کنید:

</section>

In [30]:
import pandas as pd

df = pd.read_csv('data.csv')

df["Calories"].fillna(130, inplace=True)

# print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


Replace Using Mean, Median, or Mode

<section style="direction: rtl; text-align: justify; margin: 30px;">

### **با استفاده از Mean، Median یا Mode جایگزین کنید**

یک روش رایج برای جایگزینی سلول های خالی، محاسبه میانگین، میانه یا مقدار حالت ستون است.
پانداس از متدهای mean() median() و mode() برای محاسبه مقادیر مربوطه برای یک ستون مشخص استفاده می کنند:

</section>

In [29]:
import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mean()

df["Calories"].fillna(x, inplace=True)

# print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


<section style="direction: rtl; text-align: justify; margin: 30px;">

☝🏽 میانگین = مقدار متوسط ​​(مجموع همه مقادیر تقسیم بر تعداد مقادیر).

</section>

In [27]:
import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].median()

df["Calories"].fillna(x, inplace=True)

# print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


<section style="direction: rtl; text-align: justify; margin: 30px;">

☝🏽 میانه = مقدار در وسط، بعد از اینکه همه مقادیر را به صورت صعودی مرتب کردید.

</section>

In [49]:
import pandas as pd

df = pd.read_csv('data.csv')

x = df["Calories"].mode()[0]

df["Calories"].fillna(x, inplace=True)

print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


<section style="direction: rtl; text-align: justify; margin: 30px;">

☝🏽 Mode = مقداری که بیشتر ظاهر می شود.

</section>

___
Pandas - Cleaning Data of Wrong Format

<section style="direction: rtl; text-align: justify; margin: 30px;">

## **پاکسازی داده‌های با فرمت اشتباه**

سلول هایی با داده هایی با فرمت اشتباه می توانند تجزیه و تحلیل داده ها را دشوار یا حتی غیرممکن کنند.
برای رفع آن، دو گزینه دارید: ردیف‌ها را حذف کنید یا تمام سلول‌های ستون‌ها را به یک قالب تبدیل کنید.

</section>

In [10]:
import pandas as pd

df = pd.read_csv('data2.csv')

df['Date'] = pd.to_datetime(df['Date'])

print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

<section style="direction: rtl; text-align: justify; margin: 30px;">

نتیجه تبدیل در مثال بالا به ما یک مقدار NaT داد که می‌توان آن را به عنوان یک مقدار NULL استفاده کرد و می‌توانیم با استفاده از روش dropna() ردیف را حذف کنیم.

</section>

In [11]:
df.dropna(subset=['Date'], inplace= True)